src/board/v3/nap/nap_common.c

```diff
@@ -10,19 +10,112 @@
  * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
  */
 
-#include "nap_common.h"
+#include "board.h"
+#include "nap/nap_common.h"
 
-#include "nap_regs.h"
+#include "nap_hw.h"
 #include "nap_constants.h"
 #include "axi_dma.h"
 
+#include "track.h"
+#include "system_monitor.h"
+
 #include <math.h>
 
+#define PROCESS_PERIOD_ms (1000)
+
+void nap_isr(void *context);
+
+static BSEMAPHORE_DECL(nap_exti_sem, TRUE);
+static WORKING_AREA_CCM(wa_nap_exti, 2000);
+static void nap_exti_thread(void *arg);
+
+u8 nap_track_n_channels = 1;
+
 void nap_setup(void)
 {
+  nap_track_n_channels = (NAP->STATUS & NAP_STATUS_TRACKING_CH_Msk) >>
+                          NAP_STATUS_TRACKING_CH_Pos;
+  nap_track_n_channels = MIN(nap_track_n_channels, NAP_MAX_N_TRACK_CHANNELS);
+
   axi_dma_init();
   axi_dma_start(&AXIDMADriver1);
 
   NAP->FE_PINC[0] = (u32)round(14.58e6 * pow(2.0, 32.0)
                                    / NAP_FRONTEND_SAMPLE_RATE_Hz);
+
+  /* Enable NAP interrupt */
+  chThdCreateStatic(wa_nap_exti, sizeof(wa_nap_exti), HIGHPRIO-1, nap_exti_thread, NULL);
+  gic_handler_register(IRQ_ID_NAP_TRACK, nap_isr, NULL);
+  gic_irq_sensitivity_set(IRQ_ID_NAP_TRACK, IRQ_SENSITIVITY_EDGE);
+  gic_irq_priority_set(IRQ_ID_NAP_TRACK, NAP_IRQ_PRIORITY);
+  gic_irq_enable(IRQ_ID_NAP_TRACK);
+}
+
+u64 nap_timing_count(void)
+{
+  static MUTEX_DECL(timing_count_mutex);
+  static u32 rollover_count = 0;
+  static u32 prev_count = 0;
+
+  chMtxLock(&timing_count_mutex);
+
+  u32 count = NAP->TIMING_COUNT;
+
+  if (count < prev_count)
+    rollover_count++;
+
+  prev_count = count;
+
+  u64 total_count = (u64)count | ((u64)rollover_count << 32);
+
+  chMtxUnlock(&timing_count_mutex);
+  return total_count;
+}
+
+void nap_isr(void *context)
+{
+  (void)context;
+  chSysLockFromISR();
+
+  /* Wake up processing thread */
+  chBSemSignalI(&nap_exti_sem);
+
+  chSysUnlockFromISR();
 }
+
+static void handle_nap_exti(void)
+{
+  u32 irq = NAP->TRK_IRQ;
+
+  while (irq) {
+    tracking_channels_update(irq);
+    NAP->TRK_IRQ = irq;
+
+    asm("dsb");
```

Comment:
Data synchronisation buffer.  This ensures that our write to TRK_IRQ has actually been committed before we try to read it back, in-case the CPU is trying to do something smart.  I added this earlier when we were having trouble with interrupts not arriving.  It may not be required.